components/region_cache_memory_engine/src/skiplist/list.rs

```diff
@@ -0,0 +1,1523 @@
+                                guard,
+                            )
+                            .is_ok()
+                        {
+                            // Success! Decrement the reference count.
+                            n.decrement(self.inner.memory_controller.clone(), guard);
+                        } else {
+                            self.search_bound(Bound::Included(key), false, guard);
+                            break;
+                        }
+                    }
+
+                    return true;
+                }
+            }
+        }
+    }
+
+    pub fn put(&self, key: impl Into<Bytes>, value: impl Into<Bytes>) -> Result<bool, Error> {
+        let guard = &crossbeam_epoch::pin();
+        let (key, value) = (key.into(), value.into());
+        self.check_guard(guard);
+
+        unsafe {
+            // Rebind the guard to the lifetime of self. This is a bit of a
+            // hack but it allows us to return references that are not bound to
+            // the lifetime of the guard.
+            let guard = &*(guard as *const _);
+
+            let mut search;
+            // First try searching for the key.
+            // Note that the `Ord` implementation for `K` may panic during the search.
+            search = self.search_position(&key, guard);
+            if search.found.is_some() {
+                // panic!("Overwrite is not supported, {:?}", (*r).key);
+                return Ok(false);
+            }
+
+            let height = self.random_height();
+            let (node, n) = {
+                let n = Node::create_node(key, value, height, 1, &self.inner.memory_controller)?;
+                (Shared::<Node>::from(n as *const _), &*n)
+            };
+            loop {
+                // Set the lowest successor of `n` to `search.right[0]`.
+                n.tower[0].store(search.right[0], Ordering::Relaxed);
+
+                // Try installing the new node into the skip list (at level 0).
+                if search.left[0][0]
+                    .compare_exchange(
+                        search.right[0],
+                        node,
+                        Ordering::SeqCst,
+                        Ordering::SeqCst,
+                        guard,
+                    )
+                    .is_ok()
+                {
+                    break;
+                }
+
+                // We failed. Let's search for the key and try again.
+                {
+                    // Create a guard that destroys the new node in case search panics.
+                    let sg = scopeguard::guard((), |_| {
+                        Node::finalize(node.as_raw(), self.inner.memory_controller.clone())
+                    });
+                    search = self.search_position(&n.key, guard);
+                    mem::forget(sg);
+                }
+
+                if search.found.is_some() {
+                    // panic!("Overwrite is not supported, {:?}", (*r).key);
```

Comment:
Why don't we support overwrite? The original crossbeam skiplist support it.

Reply:
Overwrite can cause problem, the crossbeam implements it by first deleting than put, so during the period, the data can be deemed as delete. In addition, we have sequence number, it should not be found.